src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs

```diff
@@ -166,69 +174,78 @@ public async Task<UserViewModel> GetAsync(string userId)
+            string roleId)
         {
-            if (_usersAndRolesDbContext.Users.Any(u => u.Email == email) || string.IsNullOrWhiteSpace(email))
-                return false;
-
-            try
-            {
-                if (_usersAndRolesDbContext.UserInvites.Any(i => i.Email == email) == false)
+            return await _userService
+                .CheckCanManageAllUsers()
+                .OnSuccess<ActionResult, Unit, UserInvite>(async () =>
                 {
-                    // TODO add role selection to Invite Users UI
-                    var analystRole = await _usersAndRolesDbContext
-                        .Roles
-                        .Where(r => r.Id == roleId)
-                        .FirstAsync();
+                    if (string.IsNullOrWhiteSpace(email))
+                    {
+                        return ValidationActionResult(ValidationErrorMessages.InvalidEmailAddress);
+                    }
+
+                    if (_usersAndRolesDbContext.Users.Any(u => u.Email.ToLower() == email.ToLower()))
+                    {
+                        return ValidationActionResult(ValidationErrorMessages.UserAlreadyExists);
+                    }
+
+                    var role = await _usersAndRolesDbContext.Roles.FirstOrDefaultAsync(r => r.Id == roleId);
+                    if (role == null)
+                    {
+                        return ValidationActionResult(ValidationErrorMessages.InvalidUserRole);
+                    }
 
                     var invite = new UserInvite
                     {
-                        Email = email,
+                        Email = email.ToLower(),
                         Created = DateTime.UtcNow,
-                        CreatedBy = user,
-                        Role = analystRole
+                        CreatedBy = inviteCreatedByUser,
+                        Role = role
                     };
-
                     await _usersAndRolesDbContext.UserInvites.AddAsync(invite);
                     await _usersAndRolesDbContext.SaveChangesAsync();
-                }
-
-                if (_usersAndRolesDbContext.UserInvites.Any(i => i.Email == email && i.Accepted == false))
-                {
                     SendInviteEmail(email);
-                }
-
-                return true;
-            }
-            catch
-            {
-                return false;
-            }
+                    return invite;
+                });
         }
 
-        public async Task<bool> CancelInviteAsync(string email)
+        public async Task<Either<ActionResult, UserInvite>> CancelInviteAsync(string email)
         {
-            var invite = _usersAndRolesDbContext.UserInvites.FirstOrDefault(i => i.Email == email);
-            _usersAndRolesDbContext.UserInvites.Remove(invite);
+            return await _userService
+                .CheckCanManageAllUsers()
+                .OnSuccess<ActionResult, Unit, UserInvite>(async () =>
```

Comment:
Was this where you were forced to supply the generic types to get this to pick a correct `OnSuccess()` method?

Reply:
Yeah. Finding it hard to get my head around the OnSuccess stuff.

Reply:
It's probably the most esoteric utility code in the project :)